core/src/avm2/globals/string.rs

```diff
@@ -201,9 +529,49 @@ pub fn create_class<'gc>(mc: MutationContext<'gc, '_>) -> GcCell<'gc, Class<'gc>
     const AS3_INSTANCE_METHODS: &[(&str, NativeMethodImpl)] = &[
         ("charAt", char_at),
         ("charCodeAt", char_code_at),
+        ("concat", concat),
+        ("indexOf", index_of),
+        ("lastIndexOf", last_index_of),
+        ("match", match_s),
+        ("slice", slice),
         ("split", split),
+        ("substr", substr),
+        ("substring", substring),
     ];
     write.define_as3_builtin_instance_methods(mc, AS3_INSTANCE_METHODS);
 
+    const PUBLIC_CLASS_METHODS: &[(&str, NativeMethodImpl)] = &[("fromCharCode", from_char_code)];
+    write.define_as3_builtin_class_methods(mc, PUBLIC_CLASS_METHODS);
+
     class
 }
+
+/// Normalizes an  index parameter used in `String` functions such as `substring`.
+/// The returned index will be within the range of `[0, len]`.
+fn string_index(i: f64, len: usize) -> usize {
+    if i == f64::INFINITY {
+        len
+    } else if i < 0. {
+        0
+    } else {
+        (i as usize).min(len)
+    }
+}
+
+/// Normalizes an wrapping index parameter used in `String` functions such as `slice`.
+/// Negative values will count backwards from `len`.
+/// The returned index will be within the range of `[0, len]`.
+fn string_wrapping_index(i: f64, len: usize) -> usize {
```

Comment:
any chance this could be unified with `string_wrapping_index` in avm1?

Reply:
I don't know how inf and -inf are handled in avm1, and there are no tests for Infinities in avm1 String tests(and I don't know how to write avm1 tests). But if it works just like in avm2, then this could be unified.